adapters/YahooAdapter/YahooAdapter/Native/GADMAdapterYahooNativeAd.m

```diff
@@ -0,0 +1,345 @@
+      }
+      return _yahooVideoPlayerView;
+    }
+  }
+
+  component = [_nativeAd component:@"mainImage"];
+  if (component) {
+    id<YASNativeViewComponent> mainImageComponent = (id<YASNativeViewComponent>)component;
+    if ([mainImageComponent conformsToProtocol:@protocol(YASNativeImageComponent)]) {
+      UIImageView *mainImageView = [[UIImageView alloc] init];
+      YASErrorInfo *error =
+          [(id<YASNativeImageComponent>)mainImageComponent prepareView:mainImageView];
+      mainImageView.userInteractionEnabled = NO;
+      if (error) {
+        NSLog(@"[YahooAdapter]: Failed to render mediaView image: %@", error);
+      }
+      return mainImageView;
+    }
+  }
+
+  return nil;
+}
+
+- (CGFloat)mediaContentAspectRatio {
+  id<YASComponent> component = [_nativeAd component:@"video"];
+  if ([component conformsToProtocol:@protocol(YASNativeVideoComponent)]) {
+    id<YASNativeVideoComponent> videoComponent = (id<YASNativeVideoComponent>)component;
+    if (videoComponent.height != 0.0f) {
+      return (videoComponent.width / videoComponent.height);
+    }
+  }
+  return 0.0f;
+}
+
+- (BOOL)hasVideoContent {
+  return self.mediaView != nil;
+}
+
+- (NSDecimalNumber *)starRating {
+  NSString *ratingString = [self stringForComponent:@"rating"];
+  if (ratingString.length > 0) {
+    NSInteger stars = 0;
+    NSInteger total = 0;
+    NSScanner *scanner = [NSScanner scannerWithString:ratingString];
+
+    NSMutableCharacterSet *set = [[NSMutableCharacterSet alloc] init];
+    [set formUnionWithCharacterSet:[NSCharacterSet letterCharacterSet]];
+    [set formUnionWithCharacterSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
+    [scanner setCharactersToBeSkipped:set];
+
+    if ([scanner scanInteger:&stars] && [scanner scanInteger:&total]) {
+      return [NSDecimalNumber
+          decimalNumberWithString:[NSString stringWithFormat:@"%ld.%ld", (long)stars, (long)total]];
+    }
+  }
+  return nil;
+}
+
+- (nullable NSDictionary *)extraAssets {
+  return nil;
+}
+
+- (void)didRecordImpression {
+  [_nativeAd fireImpression];
+}
+
+- (void)didRecordClickOnAssetWithName:(GADNativeAssetIdentifier)assetName
+                                 view:(UIView *)view
+                       viewController:(UIViewController *)viewController {
+  [_nativeAd invokeDefaultAction];
+}
+
+- (void)didRenderInView:(UIView *)view
+       clickableAssetViews:(NSDictionary<GADNativeAssetIdentifier, UIView *> *)clickableAssetViews
+    nonclickableAssetViews:
+        (NSDictionary<GADNativeAssetIdentifier, UIView *> *)nonclickableAssetViews
+            viewController:(UIViewController *)viewController {
+  id<YASComponent> iconComponent = [_nativeAd component:@"iconImage"];
+  if ([iconComponent conformsToProtocol:@protocol(YASNativeImageComponent)]) {
+    UIImageView *iconView = (UIImageView *)clickableAssetViews[GADNativeIconAsset];
+    [((id<YASNativeImageComponent>)iconComponent) prepareView:iconView];
```

Comment:
I see the icon image is handled twice, once here, and once via the 
```
- (GADNativeAdImage *)icon {
  return [self imageForComponent:@"iconImage"];
}
```
Is there a reason for this? 

The icon image ends up having different behaviour from the rest of the native components, where the clicks are handled with the YahooAds, while all other components handle the clicks with the 
didRecordClickOnAssetWithName:view:viewController: call.

Reply:
This is now removed from `didRenderInView` and the component is now prepared in `imageForComponent`.